src/topology/config/mod.rs

```diff
@@ -127,19 +127,37 @@ pub enum DataType {
 
 #[typetag::serde(tag = "type")]
 pub trait SourceConfig: core::fmt::Debug {
-    fn build(
-        &self,
-        name: &str,
-        globals: &GlobalOptions,
-        shutdown: ShutdownSignal,
-        out: mpsc::Sender<Event>,
-    ) -> crate::Result<sources::Source>;
+    fn build(&self, cx: SourceContext) -> crate::Result<sources::Source>;
 
     fn output_type(&self) -> DataType;
 
     fn source_type(&self) -> &'static str;
 }
 
+#[derive(Debug, Clone)]
+pub struct SourceContext<'a> {
+    pub name: &'a str,
```

Comment:
So the reason we went with not doing this before is that this leaks quite a lot where I find with methods that isn't the case. Happy to go with what ever method. The only opinion here is that we already have a method implemented with `SinkContext` so it would be nice to keep them all the same and change them at once in another PR.

Reply:
I agree that we should have a unified style. Also I really think this is the better approach.

What is instead of charing this to methods, merging and then changing all three to the new format I create another PR that converts the sink and transform contexts to this format and merge this afterward? It looks like less work to me.

Reply:
That seems fine as long as @lukesteensen is okay with this method.

Reply:
@LucioFranco, when you say "leaks quite a lot", what problem are you referring to? Not memory leaks I assume…

Reply:
I created #2565 that changes `SinkContext` and `TransformContext` accordingly.

Reply:
@bruceg not memory leaks but type abstraction leaks. With fn you can get around this via things like `impl Trait`. The issue for me is I very much dislike using struct items throughout the codebase because they become much harder to refactor in the future. They expose actual types. In this case I don't think its as apparent but its a style choice that I'd like for us to promote throughout the codebase.

Reply:
You're right to be concerned about abstraction leaks. On the other hand, I find trivial getter/setter methods totally distasteful. Also, while exposing the types makes refactoring more work, Rust's strict typing makes it easier to be confident with the result, so it's a mixed bag.

Reply:
We should definitely have a unified style. While I was against #2565, I feel less strongly here. The difference in my mind is that `name` is just some simple immutable data while `resolver` is something with more internal complexity. Therefore, `resolver` seems worth hiding behind a simpler API while `name` does not.

That being said, I still am somewhat against simple `pub` fields like this because of the way that it leaks implementation details of the context struct. That may not matter here, and I'd be ok with an exception for something as simple as `name` since I am also very against trivial getter/setter methods.